File: Dynamic Programming/152. Maximum Product Subarray.py
class Solution:
    def maxProduct(self, nums: List[int]) -> int:
        miniMaxArray = [[0,0] for num in nums]
        # so 0 will be the max , 1 will be the min
        
        
        #set base case here 
        miniMaxArray[0][0], miniMaxArray[0][1] = nums[0],nums[0]
        for numIndex in range(1, len(miniMaxArray)):
            miniMaxArray[numIndex][0] = max(nums[numIndex] , miniMaxArray[numIndex-1][1] * nums[numIndex], nums[numIndex] * miniMaxArray[numIndex-1][0])
            miniMaxArray[numIndex][1] = min(nums[numIndex] , miniMaxArray[numIndex-1][1] * nums[numIndex], nums[numIndex] * miniMaxArray[numIndex-1][0])
            
        finalArray = [num[0] for num in miniMaxArray]
        
        return max(finalArray)

File: Dynamic Programming/152. Maximum Product Subarray.cpp
class Solution {
public:
    int maxProduct(vector<int>& nums) {
        int n = nums.size();
        vector<int> pos(n, -1);
        vector<int> neg(n, 1);

        pos[0] = nums[0];
        neg[0] = nums[0];
        for (int i = 1; i < n; i++) {
            vector<int> temp = {nums[i], nums[i] * pos[i - 1], nums[i] * neg[i - 1]};
            pos[i] = *max_element(temp.begin(), temp.end());
            neg[i] = *min_element(temp.begin(), temp.end());
        }

        return *max_element(pos.begin(), pos.end());
    }
};